File: cli/codegen/src/CodeGenerator.kt
//
// Copyright 2025, John Clark <inindev@gmail.com>. All rights reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE file in the project root for full license information.
//
import java.io.File
import java.nio.ByteBuffer
import java.nio.file.Files
import java.nio.file.attribute.PosixFilePermission.OWNER_READ
import java.nio.file.attribute.PosixFilePermission.OWNER_WRITE
import java.text.SimpleDateFormat
import java.util.Date
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.concurrent.thread
import kotlin.math.pow
import kotlin.system.exitProcess

/** Decodes a Base32 string to bytes per RFC 4648. */
fun base32Decode(base32: String): ByteArray {
    val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    var bits = 0
    var bitCount = 0
    val bytes = mutableListOf<Byte>()

    base32.uppercase().filter { it != '=' }.forEach { char ->
        val value = alphabet.indexOf(char)
        require(value != -1) { "invalid base32 character: $char" }

        bits = bits.shl(5) or value
        bitCount += 5

        if (bitCount >= 8) {
            bitCount -= 8
            bytes += (bits shr bitCount and 0xFF).toByte()
        }
    }
    return bytes.toByteArray()
}

/** Reads and validates seed.txt, returning decoded secret bytes. */
fun readSecretBytes(): ByteArray {
    val file = File("seed.txt")
    require(file.exists()) { "seed.txt not found in current directory" }

    val ownerReadWrite = setOf(OWNER_READ, OWNER_WRITE)
    val ownerReadOnly = setOf(OWNER_READ)

    val perms = Files.getPosixFilePermissions(file.toPath())
    require(perms == ownerReadWrite || perms == ownerReadOnly) {
        "seed.txt has insecure permissions (must be 600 or 400)"
    }

    val secret = file.readText().trim()
    require(secret.isNotEmpty()) { "seed.txt is empty" }
    return base32Decode(secret)
}

/** Generates a TOTP code per RFC 6238 with HMAC-SHA1. */
fun generateTotp(secretBytes: ByteArray): String {
    val timeStep = 30L
    val digits = 6
    val counter = System.currentTimeMillis() / 1000 / timeStep
    val bytes = ByteBuffer.allocate(8).putLong(counter).array()

    val hash = Mac.getInstance("HmacSHA1").run {
        init(SecretKeySpec(secretBytes, "HmacSHA1"))
        doFinal(bytes)
    }

    val offset = hash.last().toInt() and 0x0F
    val binary = (hash[offset].toInt() and 0x7F shl 24) or
            (hash[offset + 1].toInt() and 0xFF shl 16) or
            (hash[offset + 2].toInt() and 0xFF shl 8) or
            (hash[offset + 3].toInt() and 0xFF)
    return (binary % 10.0.pow(digits).toInt()).toString().padStart(digits, '0')
}

fun main() {
    val secretBytes = try {
        readSecretBytes()
    } catch (e: Exception) {
        println("error: ${e.message}")
        exitProcess(1)
    }

    var running = true
    Runtime.getRuntime().addShutdownHook(thread(start = false) {
        running = false
        println("\nshutting down gracefully...")
    })

    println("\nauthenticator codes (ctrl+c to exit)")
    var lastLength = 0
    while (running) {
        val time = Date()
        val code = generateTotp(secretBytes)
        val timeStr = SimpleDateFormat("HH:mm:ss").format(time)
        val dots = ".".repeat(((30 - (System.currentTimeMillis() / 1000 % 30).toInt() + 2) / 3).coerceAtMost(10))
        val output = "$timeStr:  ${code.take(3)} ${code.drop(3)}  $dots"
        print("\r${output.padEnd(maxOf(output.length, lastLength), ' ')}")
        lastLength = output.length
        System.out.flush()
        Thread.sleep(1000L, 0).takeUnless { running }?.let { running = false }
    }
}

File: cli/codegen/code_generator.py
#!/usr/bin/env python3
#
# generates TOTP codes from a base32 seed (seed.txt)
#
# Copyright (c) 2025, John Clark <inindev@gmail.com>
#
# Licensed under the Apache License. See LICENSE file in the project root for full license information.
#

import base64
import hmac
import os
import stat
import sys
import time
from datetime import datetime

def base32_decode(base32: str) -> bytes:
    """Decode a Base32 string to bytes per RFC 4648."""
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    bits = 0
    bit_count = 0
    bytes_list = []

    for char in base32.upper().replace("=", ""):
        try:
            value = alphabet.index(char)
        except ValueError as e:
            raise ValueError(f"Invalid base32 character: {char}") from e

        bits = (bits << 5) | value
        bit_count += 5

        if bit_count >= 8:
            bit_count -= 8
            bytes_list.append((bits >> bit_count) & 0xFF)

    return bytes(bytes_list)

def read_secret_bytes(filename: str = "seed.txt") -> bytes:
    """Read and validate seed.txt, returning decoded secret bytes."""
    if not os.path.isfile(filename):
        raise FileNotFoundError(f"{filename} not found in current directory")

    # check seet.txt file permissions
    mode = os.stat(filename).st_mode & 0o777
    valid_perms = {stat.S_IRUSR | stat.S_IWUSR, stat.S_IRUSR}  # 600 or 400
    if mode not in valid_perms:
        raise PermissionError(f"{filename} has insecure permissions (must be 600 or 400)")

    with open(filename, "r", encoding="utf-8") as f:
        secret = f.read().strip()

    if not secret:
        raise ValueError(f"{filename} is empty")

    return base32_decode(secret)

def generate_totp(secret_bytes: bytes, timestamp: float = None) -> str:
    """Generate a TOTP code per RFC 6238 with HMAC-SHA1."""
    time_step = 30
    digits = 6
    timestamp = timestamp or time.time()
    counter = int(timestamp) // time_step
    counter_bytes = counter.to_bytes(8, "big")

    hmac_obj = hmac.new(secret_bytes, counter_bytes, "sha1")
    hash_bytes = hmac_obj.digest()

    offset = hash_bytes[-1] & 0x0F
    binary = (
        (hash_bytes[offset] & 0x7F) << 24 |
        (hash_bytes[offset + 1] & 0xFF) << 16 |
        (hash_bytes[offset + 2] & 0xFF) << 8 |
        (hash_bytes[offset + 3] & 0xFF)
    )
    return f"{binary % 10**digits:06d}"

def main():
    try:
        secret_bytes = read_secret_bytes()
    except (FileNotFoundError, PermissionError, ValueError) as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    print("\nAuthenticator codes (Ctrl+C to exit)")
    try:
        last_length = 0
        while True:
            now = datetime.now()
            code = generate_totp(secret_bytes)
            time_str = now.strftime("%H:%M:%S")
            seconds_left = 30 - (int(time.time()) % 30)
            dots = "." * min((seconds_left + 2) // 3, 10)
            output = f"{time_str}:  {code[:3]} {code[3:]}  {dots}"
            print(f"\r{output:<{max(len(output), last_length)}}", end="", flush=True)
            last_length = len(output)
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nShutting down gracefully...", file=sys.stderr)
        sys.exit(0)

if __name__ == "__main__":
    main()
